src/run_program.rs

```diff
@@ -20,11 +20,39 @@ const OP_COST: Cost = 1;
 const STACK_SIZE_LIMIT: usize = 20000000;
 
 #[cfg(feature = "pre-eval")]
-pub type PreEval =
-    Box<dyn Fn(&mut Allocator, NodePtr, NodePtr) -> Result<Option<Box<PostEval>>, EvalErr>>;
+type PreEvalIndex = usize;
 
 #[cfg(feature = "pre-eval")]
-pub type PostEval = dyn Fn(Option<NodePtr>);
+/// Tell whether to call the post eval function or not, giving a reference id
+/// for the computation to pick up.
+pub enum PreEvalResult {
+    CallPostEval(PreEvalIndex),
+    Done,
+}
+
+#[cfg(feature = "pre-eval")]
+/// Implementing this trait allows an object to be notified of clvm operations
+/// being performed as they happen.
+pub trait PreEval {
```

Comment:
I wonder if there's now a better name for this, since it could potentially implement both pre and post eval. Maybe `Tracer` or `EvalLogger` or `InvocationTracker` or `CallMonitor` or something. (Thanks to ChatGPT for some of these ideas.)

`PreEval` is maybe more useful for people who know the code already, but that's maybe eight out of eight billion. :)